Add ability to generate a focused set of targets using tags

hiltonc opened the pull request:
Resolves <https://github.com/tuist/tuist/issues/6512>

### Short description 📝

This adds the ability to annotate targets with a list of metadata tags. You can then use these tags when running `tuist generate` to focus on all targets that have a given tag.

### How to test the changes locally 🧐

1. Add launch arg `generate tag:Framework2 --path "$(SRCROOT)/fixtures/ios_app_with_frameworks"` to `tuist` scheme and run from Xcode

Expected: project generated without `App`, `AppTests`, `Framework1`, and `Framework1Tests` targets. 

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

hiltonc commented on fixtures/ios_app_with_frameworks/App/Project.swift:
```diff
@@ -1,150 +0,0 @@
-import ProjectDescription
```
This file doesn't actually appear to be used, has a syntax error, and burned a lot of time until I figured out it wasn't even being used :(.

hiltonc commented:
Before getting into tests, does this API work for metadata tags @pepicrft? I also considered a dictionary of Plist values but I prefer the type safety of this.

hiltonc commented on Tests/TuistCoreTests/Graph/GraphTraversingTests.swift:
```diff
@@ -0,0 +1,225 @@
+final class GraphTraversingTests: TuistUnitTestCase {
+    func test_filterIncludedTargets_when_graph_is_empty() {
+        // Given
+        let subject = MockGraphTraversing()
```
This might be controversial... the subject is a mock because the actual code to be tested is a protocol extension. This makes it easy to configure the given.

pepicrft replied:
Does that mean that an instance is not necessary at all? If so, what do you think if we turn the function into `static`? I think that'd be more appropriate and eliminate the confusion here.

hiltonc replied:
I looked into making `filterIncludedTargets` static, but it does make use of several other instance methods on `GraphTraversing`. With `MockGraphTraversing` I can mock those for the test. I think it would be best to keep this as an instance extension method.

danieleformichelli commented on Sources/TuistKit/Generator/GeneratorFactory.swift:
```diff
@@ -95,8 +95,8 @@ public class GeneratorFactory: GeneratorFactorying {
             testPlan: testPlan,
-            includedTargets: includedTargets,
-            excludedTargets: excludedTargets
+            includedTargets: Set(includedTargets.map(TargetPattern.pattern)),
+            excludedTargets: Set(excludedTargets.map(TargetPattern.pattern))
```
What about adding a separate options `tuist generate --tagged XXX` so that it's more discoverable and less magic?

hiltonc replied:
Would you prefer to do `--tagged Framework1 Framework2` instead of `tag:Framework1 tag:Framework2`, or in addition?

danieleformichelli replied:
I would do it instead of

pepicrft replied:
I'd keep it as it's proposed since developers might have gotten familiar with the argument-based approach for specifying the targets to focus on. We can treat the arguments as "queries", which we can make very powerful. This is [an example](https://docs.vlt.sh/cli/commands/query) I recently came across with.

pepicrft commented on Sources/ProjectDescription/TargetMetadata.swift:
```diff
@@ -0,0 +1,20 @@
+    ///   - tags: The list of tags to use. Used to select focused targets when generating the project.
+    /// - Returns: Target metadata
+    public static func metadata(
+        tags: Set<String> = []
```
Even though a set is a more appropriate data structure, the initialization of arrays is more convenient `["foo"]` (vs `Set(["foo"])`:

```suggestion
        tags: [String] = []
```

pepicrft commented on Package.swift:
```diff
@@ -464,7 +464,7 @@ let package = Package(
         .package(url: "https://github.com/tuist/swift-openapi-urlsession", branch: "swift-tools-version"),
         .package(url: "https://github.com/tuist/Path", .upToNextMajor(from: "0.3.0")),
-        .package(url: "https://github.com/tuist/XcodeGraph.git", exact: "0.10.0"),
+        .package(url: "https://github.com/hiltonc/XcodeGraph.git", branch: "target-metadata"),
```
You can update this one to 0.19.3 which includes your changes.

pepicrft commented on Sources/TuistCore/Models/TargetPattern.swift:
```diff
@@ -0,0 +1,16 @@
+/// Patterns for matching against a target.
+public enum TargetPattern: Hashable {
```
What about this instead?

```suggestion
public enum TargetQuery: Hashable {
```

pepicrft replied:
What about making this one conform `ExpressibleByStringLiteral`?

pepicrft commented on Sources/TuistKit/Mappers/FocusTargetsGraphMappers.swift:
```diff
@@ -19,6 +20,8 @@ public enum FocusTargetsGraphMappersError: FatalError, Equatable {
         case let .targetsNotFound(targets):
             return "The following targets were not found: \(targets.joined(separator: ", ")). Please, make sure they exist."
+        case .noTargetsFound:
+            return "No targets were found."
```
I'd make this one more actionable:

```suggestion
            return "No targets were found. Ensure the query is valid query and matches targets in the graph."
```

pepicrft reviewed (approved):
Some minor comments @hiltonc. 
What if we change the command argument name to `query` and its description to mention that it supports `tag:` queries?

pepicrft commented on Sources/TuistKit/ProjectEditor/ProjectEditorMapper.swift:
```diff
@@ -521,7 +521,8 @@ final class ProjectEditorMapper: ProjectEditorMapping {
             dependencies: dependencies,
-            additionalFiles: additionalFilePaths.map { .file(path: $0) }
+            additionalFiles: additionalFilePaths.map { .file(path: $0) },
+            metadata: .init(tags: [])
```
Nitpick: I'd replace `.init` with `Type(` to make it easier for the compiler to resolve things.

hiltonc replied:
Updated to `.metadata()`

pepicrft reviewed (changes requested):
We are almost there @hiltonc. I'm missing some documentation in [the argument](https://docs.tuist.dev/en/cli/generate#sources) that mentions this new filtering capability.

hiltonc commented:
> What if we change the command argument name to `query` and its description to mention that it supports `tag:` queries?

Updated the command arg value name to `query` and updated the doc to mention `tag:` queries.

hiltonc commented:
@pepicrft Sorry for the delay, I was finally able to make the requested changes.

pepicrft commented:
Thanks a lot @hiltonc for making this one happen 🙏🏼

fortmarek reviewed (approved):
Awesome stuff! Would you be up to write a post at the community forum to announce this functionality and to tell folks how they can use this? 👀

hiltonc commented:
@fortmarek Sure, I can do that.

hiltonc commented:
@fortmarek we’re not actually using this functionality yet, but I have plans to after the next Tuist release. Do you want me to write something about how we intend to use it, or wait for some real world experience?

fortmarek commented:
> Sure, I can do that.

That would be _amazing_.

> Do you want me to write something about how we intend to use it, or wait for some real world experience?

It might be more insightful if you take it for a spin first in your project, so we can also wait a little bit as long as it wouldn't be weeks after the release 🙏

hiltonc commented:
@fortmarek It looks like this isn't quite working out, when I `tuist cache` I don't get the behavior I was hoping for: https://github.com/tuist/tuist/issues/7509
